Handle adding search index twice properly

Guys please take a look on this one, we got it for some users on `2.3.3`

```
java.lang.RuntimeException: Unable to create application.android.YaloApp: java.lang.ArrayIndexOutOfBoundsException: columnIndex is less than 0.
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:3279)
       at android.app.ActivityThread.access$2200(ActivityThread.java:117)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:973)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:123)
       at android.app.ActivityThread.main(ActivityThread.java:3687)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:507)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:842)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:600)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by: java.lang.ArrayIndexOutOfBoundsException: columnIndex is less than 0.
       at io.realm.internal.Table.nativeAddSearchIndex(Table.java)
       at io.realm.internal.Table.addSearchIndex(Table.java:1156)
       at .android.model.Migration.execute(Migration.java:69)
       at io.realm.Realm.migrateRealm(Realm.java:1636)
       at io.realm.Realm.create(Realm.java:569)
       at io.realm.Realm.getDefaultInstance(Realm.java:521)
       at .model.RealmManager.getInstance(RealmManager.java:42)
       at .model.CountryDao.init(CountryDao.java:19)
       at .model.CountryDao.init(CountryDao.java:15)
       at .android.YaloApp.onCreate(YaloApp.java:66)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:969)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:3276)
       at android.app.ActivityThread.access$2200(ActivityThread.java:117)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:973)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:123)
       at android.app.ActivityThread.main(ActivityThread.java:3687)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:507)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:842)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:600)
       at dalvik.system.NativeStart.main(NativeStart.java)
```

It crashes on `userTable.addSearchIndex(userTable.getColumnIndex("appId"));`

```
public class Migration implements RealmMigration {
    @Override
    public long execute(Realm realm, long version) {
        .....

        // in 0.82 version @PrimaryKey fields become indexed. Older schemas require a migration.
        if (version == 6) {
            userTable.addSearchIndex(userTable.getColumnIndex("appId"));
            version++;
        }

        return version;
    }
```

Thanks,
Yakiv

--- a/realm/src/androidTest/java/io/realm/RealmMigrationTests.java
+++ b/realm/src/androidTest/java/io/realm/RealmMigrationTests.java
@@ -9,6 +9,7 @@ import io.realm.entities.AllTypes;
 import io.realm.entities.AnnotationTypes;
 import io.realm.entities.FieldOrder;
 import io.realm.entities.NullTypes;
+import io.realm.entities.PrimaryKeyAsLong;
 import io.realm.entities.StringOnly;
 import io.realm.exceptions.RealmMigrationNeededException;
 import io.realm.internal.ColumnType;
@@ -163,6 +164,36 @@ public class RealmMigrationTests extends AndroidTestCase {
                 fail(e.getMessage());
             }
         }
+    }
+
+    // adding search index is idempotent
+    public void testAddingSearchIndexTwice() throws IOException {
+        final boolean[] didMigrate = {false};
+
+        RealmMigration migration = new RealmMigration() {
+            @Override
+            public long execute(Realm realm, long version) {
+                Table table = realm.getTable(PrimaryKeyAsLong.class);
+                long columnIndex = table.getColumnIndex("id");
+                table.addSearchIndex(columnIndex);
+                columnIndex = table.getColumnIndex("name");
+                table.convertColumnToNullable(columnIndex);
+                didMigrate[0] = true;
+                return 42;
+            }
+        };
+        TestHelper.copyRealmFromAssets(getContext(), "default-before-migration.realm", Realm.DEFAULT_REALM_NAME);
+        RealmConfiguration realmConfig = new RealmConfiguration.Builder(getContext())
+                .schemaVersion(42)
+                .schema(PrimaryKeyAsLong.class)
+                .migration(migration)
+                .build();
+        Realm.migrateRealm(realmConfig);
+        realm = Realm.getInstance(realmConfig);
+        assertEquals(42, realm.getVersion());
+        assertTrue(didMigrate[0]);
+        Table table = realm.getTable(PrimaryKeyAsLong.class);
+        assertEquals(true, table.hasSearchIndex(table.getColumnIndex("id")));
     }
 
     public void testSetAnnotations() {

--- a/realm/src/androidTest/java/io/realm/RealmMigrationTests.java
+++ b/realm/src/androidTest/java/io/realm/RealmMigrationTests.java
@@ -10,6 +10,7 @@ import io.realm.entities.AnnotationTypes;
 import io.realm.entities.FieldOrder;
 import io.realm.entities.NullTypes;
 import io.realm.entities.PrimaryKeyAsLong;
+import io.realm.entities.PrimaryKeyAsString;
 import io.realm.entities.StringOnly;
 import io.realm.exceptions.RealmMigrationNeededException;
 import io.realm.internal.ColumnType;
@@ -168,32 +169,40 @@ public class RealmMigrationTests extends AndroidTestCase {
 
     // adding search index is idempotent
     public void testAddingSearchIndexTwice() throws IOException {
-        final boolean[] didMigrate = {false};
+        Class[] classes = {PrimaryKeyAsLong.class, PrimaryKeyAsString.class};
 
-        RealmMigration migration = new RealmMigration() {
-            @Override
-            public long execute(Realm realm, long version) {
-                Table table = realm.getTable(PrimaryKeyAsLong.class);
-                long columnIndex = table.getColumnIndex("id");
-                table.addSearchIndex(columnIndex);
-                columnIndex = table.getColumnIndex("name");
-                table.convertColumnToNullable(columnIndex);
-                didMigrate[0] = true;
-                return 42;
-            }
-        };
-        TestHelper.copyRealmFromAssets(getContext(), "default-before-migration.realm", Realm.DEFAULT_REALM_NAME);
-        RealmConfiguration realmConfig = new RealmConfiguration.Builder(getContext())
-                .schemaVersion(42)
-                .schema(PrimaryKeyAsLong.class)
-                .migration(migration)
-                .build();
-        Realm.migrateRealm(realmConfig);
-        realm = Realm.getInstance(realmConfig);
-        assertEquals(42, realm.getVersion());
-        assertTrue(didMigrate[0]);
-        Table table = realm.getTable(PrimaryKeyAsLong.class);
-        assertEquals(true, table.hasSearchIndex(table.getColumnIndex("id")));
+        for (final Class clazz : classes){
+            final boolean[] didMigrate = {false};
+
+            RealmMigration migration = new RealmMigration() {
+                @Override
+                public long execute(Realm realm, long version) {
+                    Table table = realm.getTable(clazz);
+                    long columnIndex = table.getColumnIndex("id");
+                    table.addSearchIndex(columnIndex);
+                    if (clazz == PrimaryKeyAsLong.class) {
+                        columnIndex = table.getColumnIndex("name");
+                        table.convertColumnToNullable(columnIndex);
+                    }
+                    didMigrate[0] = true;
+                    return 42;
+                }
+            };
+            RealmConfiguration realmConfig = new RealmConfiguration.Builder(getContext())
+                    .schemaVersion(42)
+                    .schema(clazz)
+                    .migration(migration)
+                    .build();
+            Realm.deleteRealm(realmConfig);
+            TestHelper.copyRealmFromAssets(getContext(), "default-before-migration.realm", Realm.DEFAULT_REALM_NAME);
+            Realm.migrateRealm(realmConfig);
+            realm = Realm.getInstance(realmConfig);
+            assertEquals(42, realm.getVersion());
+            assertTrue(didMigrate[0]);
+            Table table = realm.getTable(clazz);
+            assertEquals(true, table.hasSearchIndex(table.getColumnIndex("id")));
+            realm.close();
+        }
     }
 
     public void testSetAnnotations() {

